compiler/rustc_mir_transform/src/deduce_param_attrs.rs

```diff
@@ -0,0 +1,158 @@
+//! Deduces supplementary parameter attributes from MIR.
+//!
+//! Deduced parameter attributes are those that can only be soundly determined by examining the
+//! body of the function instead of just the signature. These can be useful for optimization
+//! purposes on a best-effort basis. We compute them here and store them into the crate metadata so
+//! dependent crates can use them.
+
+use rustc_hir::def_id::DefId;
+use rustc_index::bit_set::BitSet;
+use rustc_middle::mir::visit::{MutatingUseContext, PlaceContext, Visitor};
+use rustc_middle::mir::{Body, Local, Location, Operand, Terminator, TerminatorKind, RETURN_PLACE};
+use rustc_middle::ty::{DeducedParamAttrs, ParamEnv, TyCtxt};
+use rustc_span::DUMMY_SP;
+
+/// A visitor that determines which arguments have been mutated. We can't use the mutability field
+/// on LocalDecl for this because it has no meaning post-optimization.
+struct DeduceReadOnly {
+    /// Each bit is indexed by argument number, starting at zero (so 0 corresponds to local decl
+    /// 1). The bit is true if the argument may have been mutated or false if we know it hasn't
+    /// been up to the point we're at.
+    mutable_args: BitSet<usize>,
+}
+
+impl DeduceReadOnly {
+    /// Returns a new DeduceReadOnly instance.
+    fn new(arg_count: usize) -> Self {
+        Self { mutable_args: BitSet::new_empty(arg_count) }
+    }
+}
+
+impl<'tcx> Visitor<'tcx> for DeduceReadOnly {
+    fn visit_local(&mut self, local: Local, context: PlaceContext, _: Location) {
+        // We're only interested in arguments.
+        if local == RETURN_PLACE || local.index() > self.mutable_args.domain_size() {
+            return;
+        }
+
+        match context {
+            PlaceContext::NonMutatingUse(..) | PlaceContext::NonUse(..) => {
+                // Not mutating, so it's fine.
+                //
+                // For moves of call arguments, we converted them to synthetic mutating uses in
+                // `visit_terminator()` below.
+            }
+            PlaceContext::MutatingUse(..) => {
+                // This is a mutation, so mark it as such.
+                self.mutable_args.insert(local.index() - 1);
+            }
+        }
+    }
+
+    fn visit_terminator(&mut self, terminator: &Terminator<'tcx>, location: Location) {
+        match terminator.kind {
+            TerminatorKind::Call { ref args, ref func, ref destination, .. } => {
```

Comment:
It seems wrong to me to look at function calls. This is about attributes on the current function. Function calls within this function should have, not no effect on whether an argument is read only. In your example, we want to prevent `f`'s `x` from getting mutated by `g`, while keeping the readonly attribute on `f`'s `x` and fwiw we even keep it on `g`'s first argument, that gets then moved into `g`'s `x`

Reply:
Yes, I thought that it was strange to be looking at calls made by this function too, but @tmiasko pointed out that moves in function calls are effectively special: they can write to the memory that got moved from, and therefore have to be considered writes. No other moves in MIR behave this way.

Not taking a position on whether this is a good thing, but it's something I needed to account for.

Note that I believe it's actually impossible to observe this behavior in practice, as MIR generation always works around it by moving to a temporary first.

Reply:
> Note that I believe it's actually impossible to observe this behavior in practice, as MIR generation always works around it by moving to a temporary first.

Unsized arguments are the exception. In that case no extra copy is made.

Reply:
Ah, I see. I should update the comment then.

Reply:
In any case, the logic should be correct for unsized arguments now, as it will detect that no copy is made and disable readonly in that case. Perhaps there should be a codegen test for this.

Reply:
> Note that I believe it's actually impossible to observe this behavior in practice, as MIR generation always works around it by moving to a temporary first.

If that were true, your PR would just always return false for all entries in the slice. You are looking at optimized MIR which can end up with direct accesses to arguments instead of just the moved out version

Also I believe even without optimizations a `fn foo(ref mut x: Type)` would directly reference the argument even right after MIR building

Reply:
Ah.. I may have misunderstood the entire premise here. We're pessimizing calls to `f` to make calls to `g` within `f` sound.

I did not get this out of the comment on the first read. Please state that explicitly, and also add tests for it.

Reply:
> If that were true, your PR would just always return false for all entries in the slice. You are looking at optimized MIR which can end up with direct accesses to arguments instead of just the moved out version

If this comment is still relevant, I don't believe it can. Do you have an example? The MIR you posted doesn't have a direct move of any argument to a call in it.

> Also I believe even without optimizations a fn foo(ref mut x: Type) would directly reference the argument even right after MIR building

This doesn't seem relevant, because we're looking at calls within the function, not the signature of the function. (I think this is what your last comment means.)

> Ah.. I may have misunderstood the entire premise here. We're pessimizing calls to f to make calls to g within f sound.

Yes.

> I did not get this out of the comment on the first read. Please state that explicitly, and also add tests for it.

OK, I can reword the comment. However, I don't know how to write a test: see below.